src/charm.py

```diff
@@ -483,6 +497,104 @@ def _on_generate_self_signed_certificate_action(self, event: ActionEvent) -> Non
             }
         )
 
+    def _on_get_all_certificate_requests_action(self, event: ActionEvent) -> None:
+        """Returns all certificate requests.
+
+        Args:
+            event: Juju event.
+
+        Returns:
+            None
+        """
+        event.set_results(
+            {"Result": self.tls_certificates.get_requirer_units_csrs_with_no_certs()}
+        )
+
+    def _on_get_certificate_request_action(self, event: ActionEvent) -> None:
+        """Returns certificate request for a specific relation.
+
+        Args:
+            event: Juju event.
+
+        Returns:
+            None
+        """
+        event.set_results(
+            {
+                "Result": self.tls_certificates.get_requirer_csrs_by_unit(
+                    relation_id=event.params["relation-id"]
+                )
+            }
+        )
+
+    def _on_provide_certificate_action(self, event: ActionEvent) -> None:
+        """Provides certificate to a specific requirer unit.
+
+        Args:
+            event: Juju event.
+
+        Returns:
+            None
+        """
+        if not self.unit.is_leader():
+            event.fail(message="Action cannot be run on non-leader unit.")
+            return
+
+        if not self._action_certificates_are_valid(event):
+            event.fail(message="Action input is not valid.")
+            return
+
```

Comment:
I added a function that does that in the library and it will validate what is returned.
We can refactor setting the certificates to use it as well, however that would be out of the scope of this PR, or maybe not?

Reply:
It is added int [this PR](https://github.com/canonical/tls-certificates-interface/pull/62)
`csr_matches_certificate`